File: python/examples/demo_compare_recursive.py
"""
Demo: simultaneous smoother vs textbook RTS, agreement to machine precision.
"""

# simksmoother --- a simultaneous Kalman smoother in sparse linear algebra form.
# Copyright (C) 2026 Mico Mrkaic.
#
# Produced under the guidance and direction of Mico Mrkaic, with the
# assistance of AI (Claude, Anthropic).
#
# This file is part of the simksmoother package.
#
# simksmoother is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the LICENSE file in the package root,
# or <https://www.gnu.org/licenses/>.
#

import numpy as np
from _plot_helpers import plt, save_and_close
import simksmoother as sks

def main(seed=2):
    rng = np.random.default_rng(seed)

    n = 3
    T = 80
    F = np.array([[0.8, 0.1, 0.0],
                  [0.0, 0.7, 0.2],
                  [0.0, 0.0, 0.5]])
    H = np.array([[1.0, 0.2, 0.3],
                  [0.4, 1.0, 0.1]])
    Q = 0.1 * np.eye(3)
    R = 0.05 * np.eye(2)
    a0 = np.zeros(3)
    P0 = np.eye(3)

    Y, X = sks.simulate_const(T, F, H, Q, R, a0, P0, rng=rng)
    Xhat, J, h, info = sks.smooth_const(Y, F, H, Q, R, a0, P0, return_cov=True)
    XsT, PsT = sks.rts_smoother(Y, F, H, Q, R, a0, P0)

    mean_err = float(np.max(np.abs(Xhat - XsT)))
    cov_err = max(float(np.max(np.abs(info["Ptt"][t] - PsT[t]))) for t in range(T + 1))

    print("Simultaneous vs RTS reference")
    print(f"  max |Xhat - XsT| over all t  = {mean_err:.2e}")
    print(f"  max |P_tT,sim - P_tT,rts|    = {cov_err:.2e}")
    print(f'  (Both should be ~1e-15, since J X = h is just a different '
          f'sparse factorization of the same block-tridiagonal system.)')

    # --- Plot ----------------------------------------------------------
    # Top: the two paths overlaid (visually identical).
    # Bottom: the pointwise difference (machine-precision noise).
    tt = np.arange(T + 1)
    fig, axes = plt.subplots(2, 1, figsize=(9, 6))
    axes[0].plot(tt, Xhat[0], "b-",  lw=1.5, label="simultaneous (J\\h)")
    axes[0].plot(tt, XsT [0], "r--", lw=1.5, label="recursive (Kalman+RTS)")
    axes[0].set_title("Smoothed state 1: two methods overlaid")
    axes[0].set_ylabel("x_1"); axes[0].grid(alpha=0.3); axes[0].legend()

    diff = Xhat[0] - XsT[0]
    axes[1].plot(tt, diff, "k-", lw=1.0)
    axes[1].set_title(f"Pointwise difference, state 1 "
                      f"(max abs = {np.max(np.abs(diff)):.2e})")
    axes[1].set_xlabel("t"); axes[1].set_ylabel("Xsim - Xrts")
    axes[1].grid(alpha=0.3)
    fig.tight_layout()
    save_and_close(fig, "demo_compare_recursive")

if __name__ == "__main__":
    main()

File: octave/examples/demo_compare_recursive.m
% simksmoother --- a simultaneous Kalman smoother in sparse linear algebra form.
% Copyright (C) 2026 Mico Mrkaic.
%
% Produced under the guidance and direction of Mico Mrkaic, with the
% assistance of AI (Claude, Anthropic).
%
% This file is part of the simksmoother package.
%
% simksmoother is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program; see the LICENSE file in the package root,
% or <https://www.gnu.org/licenses/>.
%

% DEMO_COMPARE_RECURSIVE  Side-by-side check: simultaneous smoother vs
% textbook Kalman filter + Rauch-Tung-Striebel smoother.
%
% This demo serves two purposes:
%   (i)  it confirms the equivalence of the two formulations to numerical
%        precision;
%   (ii) the reference implementation in simks_rts_reference.m documents
%        the recursive algorithm explicitly, so that readers familiar with
%        the textbook presentation can match it line-by-line to the
%        sparse block-tridiagonal LDL' factorization derived in the note.

% (script: cleared and cleared screen removed for demo_main compatibility)
% Add the simksmoother/ source directory to the path, relative to where
% this demo file lives.  Works regardless of the user's CWD.
demo_dir = fileparts(mfilename('fullpath'));
oct_dir  = fullfile(demo_dir, '..', 'simksmoother');
if isfolder(oct_dir), addpath(oct_dir); end

randn('seed', 7);

T  = 200;
F  = [0.8 0.1; 0.0 0.7];
H  = [1.0 0.0; 0.0 1.0];
Q  = [0.05 0.01; 0.01 0.04];
R  = [0.20 0.00; 0.00 0.15];
a0 = [0;0];
P0 = eye(2);

[Y, Xtrue] = simks_simulate_const(T, F, H, Q, R, a0, P0);

% Simultaneous smoother with posterior covariance.
sopts.return_cov = true;
[Xsim, ~, ~, info_sim] = simks_smooth_const(Y, F, H, Q, R, a0, P0, sopts);

% Classical Kalman filter + RTS smoother (reference implementation).
[Xrts, Prts] = simks_rts_reference(Y, F, H, Q, R, a0, P0);

% Compare means.
err_mean = max(max(abs(Xsim - Xrts)));

% Compare diagonals of posterior covariance for state 1.
diag_sim = zeros(1, T+1);
diag_rts = zeros(1, T+1);
for t = 0:T
  diag_sim(t+1) = info_sim.Ptt{t+1}(1,1);
  diag_rts(t+1) = Prts{t+1}(1,1);
end
err_cov = max(abs(diag_sim - diag_rts));

fprintf('Simultaneous smoother vs textbook RTS, T = %d\n', T);
fprintf('max |Xsim - Xrts|              = %.3e\n', err_mean);
fprintf('max |Var_sim - Var_rts| state 1 = %.3e\n', err_cov);
if err_mean < 1e-9 && err_cov < 1e-9
  fprintf('PASS: the two methods agree to within roundoff.\n');
else
  fprintf('Discrepancy larger than roundoff -- check implementation.\n');
end
fprintf('\nThis is what is meant by "the recursive filter is a specialized\n');
fprintf('sparse factorization of the same block-tridiagonal system".\n');

% --- Plot ----------------------------------------------------------------
% Top: the two smoothed paths on top of each other (should look identical).
% Bottom: their pointwise difference, which is at machine precision.
tt = 0:T;
fig = figure('visible', 'on', 'position', [100 100 900 600]);

subplot(2,1,1); hold on; box on;
plot(tt, Xsim(1,:), 'b-',  'linewidth', 1.5);
plot(tt, Xrts(1,:), 'r--', 'linewidth', 1.5);
legend({'simultaneous (J\\h)', 'recursive (Kalman+RTS)'}, 'location', 'best');
title('Smoothed state 1: two methods overlaid');
xlabel('t'); ylabel('x_1');

subplot(2,1,2); hold on; box on;
plot(tt, Xsim(1,:) - Xrts(1,:), 'k-', 'linewidth', 1.0);
title(sprintf('Pointwise difference, state 1  (max abs = %.2e)', ...
              max(abs(Xsim(1,:) - Xrts(1,:)))));
xlabel('t'); ylabel('Xsim - Xrts');

fig_dir = fullfile(fileparts(mfilename('fullpath')), 'figures');
if ~exist(fig_dir, 'dir'), mkdir(fig_dir); end
fig_path = fullfile(fig_dir, 'demo_compare_recursive.png');
print(fig, fig_path, '-dpng', '-r120');
drawnow;
fprintf('Figure saved: %s\n', fig_path);
